server/src/main/java/org/opensearch/index/store/RemoteDirectory.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.opensearch.common.blobstore.BlobContainer;
+import org.opensearch.common.blobstore.BlobMetadata;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.NoSuchFileException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A {@code RemoteDirectory} provides an abstraction layer for storing a list of files to a remote store.
+ * A remoteDirectory contains only files (no sub-folder hierarchy). This class does not support all the methods in
+ * the Directory interface. Currently, it contains implementation of methods which are used to copy files to/from
+ * the remote store. Implementation of remaining methods will be added as remote store is integrated with
+ * replication, peer recovery etc.
+ */
+public class RemoteDirectory extends Directory {
```

Comment:
> provides an abstraction layer

It's a concrete implementation that abstracts the storing of files in a remote store, no? Can we make this class final so plugins do not extend? And let's mark w/ javadoc annotation @opensearch.internal as well to document this as an internal class that will not be extended outside the core.

If there is an intent to allow third-party plugins to extend this then we need an abstract class with very thoughtful overrides. Enabling a full scale override of the directory is dangerously trappy.

Reply:
Right, we don't want third-party plugins to extend `RemoteDirectory`. I will make it final and mark it `@opensearch.internal`.